bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/TypeConverter.java

```diff
@@ -828,12 +829,12 @@ private static boolean isFloatWithinIntRange(double doubleValue) {
     public static Type resolveMatchingTypeForUnion(Object value, Type type) {
         if (value instanceof ArrayValue && ((ArrayValue) value).getType().getTag() == TypeTags.ARRAY_TAG &&
                 !isDeepConversionRequiredForArray(((ArrayValue) value).getType())) {
-            return ((ArrayValue) value).getType();
+            return TypeCreator.createArrayType(((BArrayType) ((ArrayValue) value).getType()).getElementType());
         }
 
         if (value instanceof MapValue && ((MapValue) value).getType().getTag() == TypeTags.MAP_TAG &&
                 !isDeepConversionRequiredForMap(((MapValue) value).getType())) {
-            return ((MapValue) value).getType();
+            return TypeCreator.createMapType(((BMapType) ((MapValue) value).getType()).getConstrainedType());
```

Comment:
Shouldn't the type come from the target type?

The following fails atm.

```ballerina
type Map map<anydata>;

public function main() {
    map<int> & readonly m = {a: 1, b: 2};
    map<map<int>> n = {m};

    var o = checkpanic n.cloneWithType(Map);

    anydata p = o["m"];

    if p is map<anydata> {
        p["c"] = "foo";
    }
}
```

```cmd
error: {ballerina/lang.map}InherentTypeViolation {"message":"invalid map insertion: expected value of type 'int', found 'string'"}
        at foo:main(foo.bal:12)
```

This happens even when the original value is mutable (on slalpha5 also).

Reply:
In this case what would be the type of o? AFAIU it should be map<map<int>> not map<anydata>. Based on that, type should be coming from value not target type.

As per spec When `v` is a structural value, the inherent type of the value to be constructed comes from `t`. When `t` is a union, it must be possible to determine which member of the union to use for the inherent type by following the same rules  that are used by list constructor expressions and mapping constructor expressions with the contextually expected type.

Reply:
p["c"] = 123 will work

Reply:
Type of `o` should be `Map`, which is `map<anydata>`, right? 

```
# When `v` is a structural value, the inherent type of the value to be constructed
# comes from `t`.
```
`t` here is `map<anydata>` so the inherent type is `map<anydata>`?

```
# The `cloneWithType` operation is recursively applied to each member of `v` using
# the type descriptor that the inherent type requires for that member.
```
When we get to the member `m`, the type required for that is `anydata`, so shouldn't the result be the same as calling `m.cloneWithType(anydata)`?